Sources/SwiftCoAP/SCMessage.swift

```diff
@@ -45,63 +45,77 @@ extension SCCoAPTransportLayerDelegate {
-
+    func sendCoAPMessage(_ message: SCMessage, toHost host: String, port: UInt16, token: UInt64?, delegate: SCCoAPTransportLayerDelegate?) throws
+    func sendCoAPMessage(_ message: SCMessage, toEndpoint endpoint: NWEndpoint, token: UInt64?, delegate: SCCoAPTransportLayerDelegate?) throws
+    
+    func getMessageId(for endpoint:NWEndpoint) -> UInt16
     // Called when the transmission is over. Clear your states (e.g. close sockets)
-    func closeTransmission()
+    func closeTransmission(for endpoint: NWEndpoint, withToken: UInt64)
     
-    // Start to listen for Messages. Prepare e.g. sockets for receiving data. This method will only be called by SCServer
-    func startListening() throws
-
-    // Same as `startListening()` but on non-default port.
-    func startListening(onPort listenPort: UInt16) throws
 }
 
 extension SCCoAPTransportLayerProtocol {
-    public func sendCoAPData(_ data: Data, toHost host: String, port: UInt16) throws {
-        try sendCoAPData(
-            data,
+    public func sendCoAPMessage(_ message: SCMessage, toHost host: String, port: UInt16, token: UInt64?, delegate: SCCoAPTransportLayerDelegate?) throws {
+        try sendCoAPMessage(
+            message,
             toEndpoint: NWEndpoint.hostPort(
                 host: NWEndpoint.Host(host),
                 port: NWEndpoint.Port(rawValue: port)!
-            )
+            ),
+            token: token,
+            delegate: delegate
         )
     }
 }
 
 
+public struct MessageTransportIdentifier: Hashable {
+    let token: UInt64;
+    let endpoint: NWEndpoint
+}
+
+public struct MessageTransportDelegate {
+    let delegate: SCCoAPTransportLayerDelegate;
+    let observation: Bool
+}
+
+public struct CoAPConnection {
+    let connection: NWConnection;
+    var lastReceivedMessageTs: TimeInterval;
+    var pingTimer: Timer?;
+}
 
 //MARK: - SC CoAP UDP Transport Layer
 /// SC CoAP UDP Transport Layer: This class is the default transport layer handler, sending data via UDP with help of `Network.framework`. If you want to create a custom transport layer handler, you have to create a custom class and adopt the SCCoAPTransportLayerProtocol. Next you have to pass your class to the init method of SCClient: init(delegate: SCClientDelegate?, transportLayerObject: SCCoAPTransportLayerProtocol). You will than get callbacks to send CoAP data and have to inform your delegate (in this case an object of type SCClient) when you receive a response by using the callbacks from SCCoAPTransportLayerDelegate.
 public final class SCCoAPUDPTransportLayer: NSObject {
-    weak public var transportLayerDelegate: SCCoAPTransportLayerDelegate!
-    var connections: [NWEndpoint: NWConnection] = [:]
-    var listener: NWListener?
-    var networkParameters: NWParameters = .udp
+    fileprivate let kPingInterval:TimeInterval = 5
+    fileprivate var transportLayerDelegates: [MessageTransportIdentifier: MessageTransportDelegate] = [:]
```

Comment:
Have to check thoroughly. Won't it lead to reference cycle between the transport and its delegates? I have rejected the exact same idea when I was thinking about the possible solution last Nov.

Reply:
it is valid point, if client observing the resource is dropped without calling `cancelMessageTransmission` delegate will not be removed from the dictionary and will stay allocated.

I also added cleanup code for the delegates to be released when connection is canceled to the endpoint (or message sending failed)